src/attr/__init__.pyi

```diff
@@ -86,6 +86,16 @@ else:
         takes_self: bool = ...,
     ) -> _T: ...
 
+# Static type inference support via __dataclass_transform__ implemented as per:
+# https://github.com/microsoft/pyright/blob/1.1.135/specs/dataclass_transforms.md
+# This annotation must be applied to all overloads of "define" and "attrs"
+def __dataclass_transform__(
```

Comment:
There should probably be an implementation of this somewhere right?  Otherwise this won't be useful outside of attrs.

Reply:
This is a little strange, the decorator is just being used as a marker for `pyright`. The upstream spec specifies that the `__dataclass_transform__` decorator is defined on a per-project basis _either_ in the a `.py` or `.pyi` source file, essentially as a compatibility layer until this is moved into `typing_extensions`.

Reply:
Hmm.  I'm a little concerned (ok not too much but it definitely feels weird) about having a `def ` in the pyi file that doesn't exist in the .py file.  It means that someone could try to use this function in their own code but mypy wouldn't warn them that it doesn't really exist.  It would fail at runtime.  Are we sure we don't want to put an implementation in _funcs.py or something?

Reply:
🤔 I've done a little test of this. Under the 1.1.135 `pyright` implementation if we:

* Define a `__dataclass_transform__` decorator in `__init__.pyi` file.
* Define a `__dataclass_transform__` method in `__funcs.py` and import in `__init__.py`.

Then a custom annotation of the form has types inferred properly.

```python
import attr

@attr.__dataclass_transform__(fields=(attr.field, attr.attrib))
def custom_define(cls):
   return attr.define(cls)

@custom_define
class Custom:
    a: int
```

*However*, given that this is a very early and provisional specification and implementation in `pyright`, adding a "surprisingly functional" shim in `attrs`  introduces a potentially risky coupling. We've a choice between:

1. An immediate and loud runtime failure if you "try to use" the attrs-internal `__dataclass_transform__` `.pyi` definition, that won't be caught static type checkers.
2. A surprisingly functional form that "works" (a no-op a runtime, currently works at typingtime), but relies on unspecified upstream behavior.

Given how bleeding-edge this feature is I have a weakly held preference that we opt for 2 for this release to ensure that folks don't implictly rely on the `import attr.__dataclass_transform__` behavior and instead wait until the spec bakes enough for the `dataclass_transform` to show up in `typing_extensions`.

Of course, happy to reconsider.

Reply:
Since this is only for pyright and doesn't currently get mypy to do the right thing (i.e. it doesn't tell mypy that this is an attrs class creator) perhaps it's best if we leave this only inside the pyi file.  I might add a comment on here that says that this method doesn't actually exist at runtime.

Reply:
Roger, comment added.